Make bindings discoverable on system startup

andrewfg opened the pull request:
In relation to the PRs listed below, this PR adds discovery method XML to 70 existing bindings, and thus makes them discoverable on system startup.

- https://github.com/openhab/openhab-core/pull/3806
- https://github.com/openhab/openhab-core/pull/3865

**Addons with Issues**

There are some bindings where issues have been identified, so these have not yet been included in this PR.
- 'tapocontrol' binding

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

hmerk reviewed (commented):
I think you can exclude the Wemo Binding, unless you have implemented the specific upnp discovery used in the binding. Wemo devices in the past did not answer to standard upnp discovery.

andrewfg commented:
> exclude the Wemo Binding, unless you have implemented the specific upnp discovery used in the binding. 

@hmerk the purpose is to _suggest_ potential bindings to install; so as a general rule I tried discover devices on the network that have a high suspicion of being related to a particular addon, rather than trying to check all details.

In the case of wemo it would look for upnp devices that report a manufacturer name "BELKIN"..

mlobstein reviewed (approved):
LGTM

mherwege commented on bundles/org.openhab.binding.upnpcontrol/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -15,4 +15,26 @@
+			</match-properties>
+		</discovery-method>
+		<discovery-method>
+			<service-type>upnp</service-type>
```
You can remove discovery for MediaServer. Without a media renderer from the binding, it is useless. The media renderer on its own can be used as an audio sink. So proposing the binding if there only is a server makes no sense.

andrewfg replied:
Umm. In my home the server is mains powered, but the renderers are battery powered.

mherwege replied:
Ok, fair enough.

hmerk commented:
> > exclude the Wemo Binding, unless you have implemented the specific upnp discovery used in the binding.
> 
> @hmerk the purpose is to _suggest_ potential bindings to install; so as a general rule I tried discover devices on the network that have a high suspicion of being related to a particular addon, rather than trying to check all details.
> 
> In the case of wemo it would look for upnp devices that report a manufacturer name "BELKIN"..

I understood this, but it is useless, as they will not show up on a standard upnp discovery. Belkin has blocked this way of discovery.

mherwege commented on bundles/org.openhab.binding.neeo/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -10,4 +10,11 @@
 		on the Devices</description>
 	<connection>local</connection>
 
+	<discovery-methods>
```
The neeo transport could be discovered in the same way.

andrewfg replied:
Ok. Done.

andrewfg commented:
> Belkin has blocked this way of discovery, so we will never see a suggestion to install the Wemo binding.

Ok. I can easily delete the Wemo entry. But @hmerk your point does raise the bigger question of why the binding still has a UpnpDiscoveryParticipant class in it? So should that class be deleted?

hmerk commented:
> > Belkin has blocked this way of discovery, so we will never see a suggestion to install the Wemo binding.
> 
> Ok. I can easily delete the Wemo entry. But @hmerk your point does raise the bigger question of why the binding still has a UpnpDiscoveryParticipant class in it? So should that class be deleted?

No, it needed to be implemented  to be able to do the Wemo specific upnp search after binding started.
It is just they do not answer on „normal“ root-device querys. After the individual scan was triggered, UpnpDiscoveryParticipant handels the incoming messages correct.
I hope to have it explained understandable.

andrewfg commented:
> After the individual scan was triggered, UpnpDiscoveryParticipant handels the incoming messages correct. I hope to have it explained understandable.

Aha. Many thanks.

mherwege commented:
The nikohomecontrol binding could be suggested with mdns, mdns-service-type ._http._tcp, match property hostname starting with FP00112A. The binding itself does not do discovery this way, but this would identify it.

gdolfen reviewed (approved):
LGTM for enigma2

andrewfg commented:
Just a short update to everybody who is kindly contributing to this.
1. First of all many thanks for your contributions!
2. I just did a full clean OH fresh build, and the following are the results from my own OH system:
- It found and suggested 9 addons quite correctly
- It found and suggested 1 addon that it should **_not_** have (one of the filters is probably not specific enough)
- It failed to find 2 addons that it **_should_** have (one of the filters is probably too specific)

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -7,4 +7,18 @@
+			<mdns-service-type>_http._tcp.local.</mdns-service-type>
+			<match-properties>
+				<match-property>
+					<name>name</name>
```
I think it should be hostname, but didn’t test.

andrewfg replied:
I think if the fully qualified name is `foobar._http._tcp.local.` then the name part is`foobar`. (name.application.transport.domain). So when you speak of `hostName` are you talking about something other than `foobar`?

mherwege replied:
Not sure. I took it from their documentation, and there it is called hostname. I guess I will have to try and confirm what the right pattern is.

mherwege replied:
I tested and this is the correct version:
```
    "discoveryMethods": [
      {
        "serviceType": "mdns",
        "mdnsServiceType": "_http._tcp.local.",
        "matchProperties": [
          {
            "name": "hostname",
            "regex": "^(?i)FP00112A"
          }
        ]
      }
    ]
```

andrewfg replied:
It seems that in this case 'hostname' is not actually referring to the _"host name"_ of the device according to standard network parlance i.e. the `foobar` part of `foobar._http._tcp.local.` .. but rather, it appears to be device specific custom property called (confusingly) 'hostname'. => I will update the PR.

andrewfg replied:
Actually I think your `<regex>^(?i)FP00112A</regex>` is wrong (it matches almost anything). Probably should be `<regex>(?i)FP00112A.*</regex>` => ??

andrewfg replied:
On my system -- where there is definitely NOT a niko device -- with your code it was suggesting me to install that binding.

mherwege replied:
I think there is a bug in the core mdns finder causing this: https://github.com/openhab/openhab-core/blob/fd11a996b4b6f09daf80c723d11320ad001f2f5c/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/finders/MDNSAddonSuggestionFinder.java#L90

If the property does not exist in the service, it is not validated. You need to check all properties effectively exist. I don’t think the problem is with the regex.

mherwege replied:
So, correct version (and I now think this is final):
```
	<discovery-methods>
		<discovery-method>
			<service-type>mdns</service-type>
			<mdns-service-type>_http._tcp.local.</mdns-service-type>
			<match-properties>
				<match-property>
					<name>name</name>
					<regex>(?i)FP00112A.*</regex>
				</match-property>
			</match-properties>
		</discovery-method>
	</discovery-methods>
```

andrewfg replied:
Yup. I already committed that last night.

mherwege replied:
Your commit still has `hostname` instead of `name` for the property.

mherwege commented:
Great to hear that.

> I just did a full clean OH fresh build

Any chance to make your full build available somewhere? I will then try my UI PR on that build.

hmerk reviewed (approved):
LGTM

andrewfg commented:
> Any chance to make your full build available somewhere?

Tomorrow. i.e. actually later today..

andrewfg commented:
> Any chance to make your full build available somewhere?

@mherwege => https://github.com/andrewfg/openhab-core/releases/tag/test

andrewfg commented:
> Any chance to make your full build available somewhere? 

@mherwege => https://github.com/andrewfg/openhab-core/releases/edit/test

david-pace commented on bundles/org.openhab.binding.boschshc/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -8,4 +8,11 @@
+	<discovery-methods>
+		<discovery-method>
+			<service-type>mdns</service-type>
+			<mdns-service-type>_homeconnect._tcp.local.</mdns-service-type>
```
Where can I find more info what `_homeconnect` means? Where is this identifier evaluated?

andrewfg replied:
Homeconnect devices seem to advertise themselves via mDNS with the _homeconnect application on the _tcp transport an the local domain..

![image](https://github.com/openhab/openhab-addons/assets/893994/3207fc00-dae6-4098-aab5-a6c0037376d9)

maniac103 replied:
But Homeconnect devices are handled by the `homeconnect` binding, not the `boschshc` one?

andrewfg replied:
You are right. Good catch. I will change it.

david-pace replied:
The Bosch SHC binding implements a generic mDNS lookup with this pattern: `_http._tcp.local.`. It then tries to reach a specific URL on the provided hosts to check whether it is a Bosch Smart Home Controller. I'm not sure if `_http._tcp.local.` would be the correct pattern in this case or if this would basically suggest the binding to all users (which would probably not be intended).

andrewfg replied:
> not sure if _http._tcp.local. would be the correct pattern in this case or if this would basically suggest the binding to all users

Indeed. The `_http._tcp.local.` application is very generic and that alone would result in the binding being falsely suggested too often. So we need to apply an additional more exclusive test. The approach used by Bosch SHC of "call home and check credentials" is a model used also by other addons, so I am aware of it. However it is probably not something that we could or should implement in OH core, since it is probably not sufficiently manufacturer agnostic. Ideally we should try to see if there is an mDNS custom attribute within the `_http._tcp.local.` response that is a specific for Bosch SHC. => Perhaps you could run an mDNS browser test on your LAN to see if there are some such attributes analog to the example of  another `_http._tcp.local.` on my own LAN (as shown below).

![image](https://github.com/openhab/openhab-addons/assets/893994/a4b5936a-851c-4a9d-afcf-1a605639343f)

andrewfg replied:
I just did the following..
- copied the addon.xml entries from BoschSHC over to HomeConnect => so the suggestion finder would now properly suggest that addon
- did NOT (yet) delete the entries from BoschSHC .. pending feedback from @david-pace => so the suggestion finder would still wrongly suggest that addon; and if @david-pace cannot find a precise search marker in the mDNS then I will finally delete the entries for this addon

david-pace replied:
@andrewfg I found the same app you are using, and searched for `_http._tcp`. I see the Bosch Smart Home Controller but it only shows

* IP Address (it is not the local network address, but a `169.254.*` address
* Port: `8443`
* Service Type: `_http._tcp`

and no attributes. Do I have to configure something to see attributes, or does this mean that the controller has no specific attributes?

andrewfg replied:
@david-pace thankyou for checking this. Unfortunately it seems that the info from the mdns discovery does not have any markers to exclusively identify the bosch shc eco system. So I shall have to exclude this one (at least for the time being) from this auto suggestion process.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -7,4 +7,18 @@
+			<mdns-service-type>_http._tcp.local.</mdns-service-type>
+			<match-properties>
+				<match-property>
+					<name>hostname</name>
```
Testing again with TRACE:
```
09:28:27.620 [TRACE] [on.finders.MDNSAddonSuggestionFinder] - mDNS service name=FP00112A22751D._http._tcp.local., properties=[]
```

So it looks like the property should be `name` after all.

andrewfg replied:
If you are using my build from yesterday, I think there was a bug whereby the trace log was not showing the properties. I made a fix of the logging last night, but I need to test again.

mherwege replied:
I am running in Eclipse, not your build from yesterday. I pulled updates of all PR's this morning.

andrewfg replied:
So the logging is apparently still not working. Or did you see other log entries where the properties list IS populated?

mherwege replied:
I also so this:
```
09:28:28.160 [TRACE] [on.finders.MDNSAddonSuggestionFinder] - mDNS service name=SCC-00000874._http._tcp.local., properties=[SCC-00000874._http._tcp.local.=true]
09:28:28.161 [TRACE] [on.finders.MDNSAddonSuggestionFinder] - mDNS service name=HP Color LaserJet MFP M281fdw (0CACA4)._http._tcp.local., properties=[HP Color LaserJet MFP M281fdw (0CACA4)._http._tcp.local.=true]
```

andrewfg replied:
Yes the properties log listing is still not showing the real actual properties. I am working on it..

andrewfg replied:
There seems to be a thread synchronization issue with the JMDNS getPropertyNames Enumeration (and/or the parallelism of stream operations) which makes the logging difficult. I am really struggling with it. So I may just remove some of the logging again.

stefan-hoehn commented:
to be honest, I have no clue what the impact of your change is, so it is hard for me to say it is fine or not. Does it mean I have to build my addon based on your changes, deploy it and see if the scan still works?

hmerk commented:
@stefan-hoehn 
No, those changes have no impact to your binding. It just adds some information that will be used in the new addon suggestion.

andrewfg commented:
> Does it mean I have to build my addon based on your changes, deploy it and see if the scan still works?

@stefan-hoehn No. I don't know which addon you are referring to, but probably it implements either a UPnPDiscoveryParticipant or an MDNSDiscoveryParticipant for discovering things. If so those classes will be some syntax for checking if a returned UPnP RemoteDevice resp. a returned MDNS ServiceInfo contains data fields that match you addon. So can your please just confirm if the changes in this PR to the `addon.xml` file for you resp. addon are consistent with the resp. syntax checks in your xyzDiscoveryParticipant(s)?

stefan-hoehn commented:
Sorry, I should have mentioned the Addon. It is the Nanoleaf-Binding and the discovery takes place in class _NanoleafMDNSDiscoveryParticipant implements MDNSDiscoveryParticipant_.

To answer you question 

> So can your please just confirm if the changes in this PR to the addon.xml file for you resp. addon are consistent with the resp. syntax checks in your xyzDiscoveryParticipant(s)?

You added

```
<discovery-methods>
		<discovery-method>
			<service-type>mdns</service-type>
			<mdns-service-type>_nanoleafapi._tcp.local.</mdns-service-type>
		</discovery-method>
	</discovery-methods>
```

and this is what I do during discovery:

```
   public @Nullable DiscoveryResult createResult(ServiceInfo service) {
        final ThingUID uid = getThingUID(service);
        if (uid == null) {
            return null;
        }
        final Map<String, Object> properties = new HashMap<>(2);
        String host = service.getHostAddresses()[0];
        properties.put(CONFIG_ADDRESS, host);
        int port = service.getPort();
        properties.put(CONFIG_PORT, port);
        String firmwareVersion = service.getPropertyString("srcvers");
        properties.put(Thing.PROPERTY_FIRMWARE_VERSION, firmwareVersion);
        String modelId = service.getPropertyString("md");
        properties.put(Thing.PROPERTY_MODEL_ID, modelId);
        properties.put(Thing.PROPERTY_VENDOR, "Nanoleaf");
        String qualifiedName = service.getQualifiedName();
        logger.debug("Device found: {}", qualifiedName);

        // ... other stuff

        return DiscoveryResultBuilder.create(uid).withThingType(getThingType(service)).withProperties(properties)
                .withLabel(service.getName()).withRepresentationProperty(CONFIG_ADDRESS).build();
    }
```

I am still not sure what "checks" you are referring to.

andrewfg commented:
> I am still not sure what "checks" you are referring to.

I wrote the addons.xml for your addon based on your java code (part of which you requoted in your post above) I think what I wrote in the XML is a reasonable interpretation of your java code, but I am asking if you agree with said interpretation.

stefan-hoehn reviewed (approved):
LGTM 4 Nanoleaf

lolodomo commented:
For lgwebos binding, you could use UPnP discovery and a matching of manufacturer with "LG Electronics" + model name with "LG.*TV".

andrewfg commented on bundles/org.openhab.binding.freebox/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -8,4 +8,11 @@
 	<description>The Freebox binding requests your Freebox Server for various operational informations.</description>
 	<connection>local</connection>
 
+	<discovery-methods>
+		<discovery-method>
+			<service-type>mdns</service-type>
+			<mdns-service-type>_fbx-api._tcp.local.</mdns-service-type>
+		</discovery-method>
+	</discovery-methods>
```
OK

andrewfg commented on bundles/org.openhab.binding.homeconnect/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -9,4 +9,11 @@
 		to household devices from brands like Bosch and Siemens.</description>
 	<connection>cloud</connection>
 
+	<discovery-methods>
+		<discovery-method>
+			<service-type>mdns</service-type>
+			<mdns-service-type>_homeconnect._tcp.local.</mdns-service-type>
+		</discovery-method>
+	</discovery-methods>
```
Yes. It works on my own system. But probably to be consequent we need to add the discovery also to the binding itself..

lolodomo replied:
The binding uses a cloud connection. That is certainly why there is no discovery implemented on local network.

lolodomo replied:
But if mDNS is working, that's very well for your purpose.

andrewfg replied:
Indeed. As you say the binding cannot discover things unless they are in the cloud. But the devices that would be in the cloud do advertise themselves via mDNS locally. So it is reasonable to **_suggest_** to install that binding..

lolodomo replied:
We agree.

lolodomo replied:
I just tried the Android app "mDNS Discovery" you are using and I can confirm that my washing machine exposed itself with _homeconnect._tcp

andrewfg commented on bundles/org.openhab.binding.sonos/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -19,4 +19,16 @@
+			<match-properties>
+				<match-property>
+					<name>manufacturer</name>
+					<regex>(?i).*SONOS.*</regex>
```
The (?i) means case insensitive..

lolodomo replied:
So it should be OK

andrewfg commented:
> matching of manufacturer with "LG Electronics" + model name with "LG.*TV".

@lolodomo may I suggest modelName to match ".*TV.*" as the LG prefix is probably already implicit via the manufacturer property check?

lolodomo commented:
> > matching of manufacturer with "LG Electronics" + model name with "LG.*TV".
> 
> @lolodomo may I suggest modelName to match `.*TV.*` as the LG prefix is probably already implicit via the manufacturer property check?

OK

lolodomo commented:
I am now playing with the Android app "mDNS Discovery".

First surprise, the Sonos devices are exposing themselves through mDNS with `_sonos._tcp`
It could be a better option to use mDNS for Sonos instead of UPnP.

Second surprise, I have a Netatmo device that exposes itself. You can use `_netatmo_lcomm._tcp`

Third surprise, my LG TV also expose some services.

For binding networkupstools, you can use `_nut._tcp`

andrewfg commented:
> Sonos devices are exposing themselves through mDNS with _sonos._tcp 
> It could be an option to use mDNS for Sonos instead of UPnP.

Ok. Added.

> Netatmo device that exposes itself. You can use _netatmo_lcomm._tcp

Ok. Added.

> Third surprise, my LG TV also expose some services.

Ok. Let me know which one (or all).

> For binding networkupstools, you can use _nut._tcp

Ok. Added.

lolodomo commented:
> > Netatmo device that exposes itself. You can use _netatmo_lcomm._tcp
> 
> Ok. Added.

Sorry, it is _netatmo-lcomm._tcp and not _netatmo_lcomm._tcp

hmerk commented:
@andrewfg 
just found that the ism binding could be added too.
ISM8 device announces through mdns as

ISM8i._ism7._tcp.local.

ISM8.local:9091
192.168.0.221:9091

ISM8 controller

jlaur commented on bundles/org.openhab.binding.miele/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -8,4 +8,11 @@
 	<description>This is the binding for Miele@home appliances</description>
 	<connection>local</connection>
 
+	<discovery-methods>
+		<discovery-method>
+			<service-type>mdns</service-type>
+			<mdns-service-type>_mieleathome._tcp.local.</mdns-service-type>
+		</discovery-method>
+	</discovery-methods>
```
This will correctly identify the XGW 3000 gateway used for the old local Zigbee-based Miele@home, i.e. the **miele** binding. So looks good:

![image](https://github.com/openhab/openhab-addons/assets/19519842/01b299c6-bc2e-42dd-acd6-3f39d4022199)

However, it will also identify Miele Wi-Fi devices, such as my dishwasher:

![image](https://github.com/openhab/openhab-addons/assets/19519842/984d137d-b481-4592-b9f2-9088f870e7ef)

Now comes the tricky part: Newer Wi-Fi devices are usually integrated through Miele Cloud. But XGW 3000 is also integrated with Miele Cloud so that the Miele app is compatible with both Wi-Fi and Zigbee communication modules.

This means that the **mielecloud** binding can be used to integrate all my appliances - Zigbee/XGW 3000 as well as Wi-Fi/cloud. And the "miele" binding can also be used to integrate all my appliances, also the Wi-Fi/cloud-based ones.

Because of this, I think the best we can do is to suggest both **miele** and **mielecloud** based on the same discovery-method definition. Unless the name "Miele XGW3000" can be matched as well? In this case this could be used to match **miele** and everything else could be used to match **mielecloud**. That would mean that if **only** XGW 3000 is found, but no other devices, **mielecloud** will not be suggested. I think this is okay, since users not having **any** Wi-Fi devices would probably prefer only the local integration anyway.

andrewfg replied:
Umm.

IMHO I think we should suggest both addons when more that one might apply. I did this already for 'hpprinter' and 'ipp' printer, and for MediaRenders like Sony, Yamaha, LG etc. which may both recommended the manufacturer specific addons and also the generic `upnpcontrol` addon.

But in the Miele case we can probably be more specific
- service type `_mieleathome._tcp.local.` causes `mielecloud` to (always) be suggested
- service type `_mieleathome._tcp.local.` combined with `path=\Rest` causes `miele` to be suggested too

Or ??

jlaur replied:
Is it possible to match by the absence of "/rest/"? That would be even better for matching **mielecloud**.

andrewfg replied:
> Is it possible to match by the absence of "/Rest/"?

The complete absence of a 'path' property would cause the pattern matcher to fail (i.e. 'null' property). But fortunately your two above mentioned mDNS results have either 'path=\' or 'path=\Rest\', so we could probably discern on that. For example as follows..

**Cloud**
```
	<discovery-methods>
		<discovery-method>
			<service-type>mdns</service-type>
			<mdns-service-type>_mieleathome._tcp.local.</mdns-service-type>
			<match-properties>
				<match-property>
					<name>path</name>
					<regex>\/</regex>
				</match-property>
			</match-properties>
		</discovery-method>
	</discovery-methods>
```

**Gateway**
```
	<discovery-methods>
		<discovery-method>
			<service-type>mdns</service-type>
			<mdns-service-type>_mieleathome._tcp.local.</mdns-service-type>
			<match-properties>
				<match-property>
					<name>path</name>
					<regex>\/Rest\/</regex>
				</match-property>
			</match-properties>
		</discovery-method>
	</discovery-methods>
```

lolodomo commented:
For the remote openhab binding, it should be suggested only if the server exposing itself is not the current server. 
If not, this binding will be suggested to everyone.

andrewfg commented:
> ISM8 device announces through mdns as
> ISM8i._ism7._tcp.local.

@hmerk I am guessing that `ISM8i._ism7._tcp.local.` is the fully qualified name of a specific device instance and so `_ism7._tcp.local.` would be the mdns service **type** to scan for

andrewfg commented:
> remote openhab binding, it should be suggested only if the server exposing itself is not the current server.
If not, this binding will be suggested to everyone.

Good point. To exclude such an "own" would require checking if `mdns-discovered-ip-address == own-ip-address`. I could indeed build such a check into the finder service at service level (rather in the binding addon.xml) but such a check would not only exclude an an own OH remote server, but would also exclude any other mDNS discovered 'device' that are running on the same PC as OH. I don't know if OH currently has 'device's that run on itself, so I don't know if such check pose a real problem. => WDYT?

jlaur commented:
> but would also exclude _ANY_ other mDNS discovered 'device' that is running on the same PC as OH

Some users have deCONZ running on the same machine as openHAB. It's probably also quite common to have mosqitto running on the same machine, but I don't think we have MQTT broker discovery yet (as it would need some more discovery methods).

ccutrer commented:
I'd caution against that. I run _lots_ of automation related things on the same Linux server as openHAB. Currently none of them use mDNS, but an example would be if an Apple HomeKit binding (to control HomeKit devices, not to expose openHAB to HomeKit controllers) were to be written, that binding should be suggested when Homebridge is being run on the same server. Though the argument could be made that for Homebridge the most common configuration is to use Docker, and indeed that many people run openHAB in Docker, and that anyone running multiple services on the same server often use Docker. But that's not a guarantee.

lolodomo commented:
> . I could indeed build such a check into the finder service at service level (rather in the binding addon.xml) but such a check would not only exclude an own OH remote server, but would also exclude _ANY_ other mDNS discovered 'device' that is running on the same PC as OH. I don't know if OH currently has 'device's that run on itself, so I don't know if such check poses a real problem. => WDYT?

Yes, it can happen, for example for UPS with the nut server running on the same server as OH.

andrewfg commented:
> mosquitto running on the same machine

Indeed. I was already thinking of adding an `MQTTAddonSuggestionFinder` service at some future stage. But not in this PR please.

> lots of automation related things on the same Linux server as openHAB

Yeah. Perhaps we could have an `exclude-local-devices` attribute in the `discovery-method` xml schema, so that `exclude-local-devices=true` causes devices with the same ip address to be excluded (but otherwise always included). We could put that attribute on the addon.xml for the special case of the OH remote server, whereas probably for most others we can leave out the such attribute. => WDYT?

lolodomo commented:
> Perhaps we could have an `exclude-local-devices` attribute in the `discovery-method` xml schema, so that `exclude-local-devices=true` causes devices with the same ip address to be excluded (but otherwise always included). We could put that attribute on the addon.xml for the special case of the OH remote server, whereas probably for most others we can leave out the such attribute. => WDYT?

Yes, I think that is the way to go.

hmerk commented:
@andrewfg 
> > ISM8 device announces through mdns as
> > ISM8i._ism7._tcp.local.
> 
> @hmerk I am guessing that `ISM8i._ism7._tcp.local.` is the fully qualified name of a specific device instance and so `_ism7._tcp.local.` would be the mdns service **type** to scan for

Yes, that’s the one.

andrewfg commented:
> Yes, that’s the one.

But interesting that the mdns service is ism7 and the binding is ism8 ..

lolodomo commented:
> > Perhaps we could have an `exclude-local-devices` attribute in the `discovery-method` xml schema, so that `exclude-local-devices=true` causes devices with the same ip address to be excluded (but otherwise always included). We could put that attribute on the addon.xml for the special case of the OH remote server, whereas probably for most others we can leave out the such attribute. => WDYT?
> 
> Yes, I think that is the way to go.

At the same time, if you are implementing something only for the remote openHAB binding and this is a lot of work for you, another option is just to not do it and to never suggest this binding...

hmerk commented:
> > Yes, that’s the one.
> 
> But interesting that the mdns service is ism7 and the binding is ism8 ..

There are also ISM7 and ISM9 devices, but don’t know how they work…

hmerk commented:
> > > Perhaps we could have an `exclude-local-devices` attribute in the `discovery-method` xml schema, so that `exclude-local-devices=true` causes devices with the same ip address to be excluded (but otherwise always included). We could put that attribute on the addon.xml for the special case of the OH remote server, whereas probably for most others we can leave out the such attribute. => WDYT?
> > 
> > 
> > Yes, I think that is the way to go.
> 
> At the same time, if you are implementing something only for the remote openHAB binding and this is a lot of work for you, another option is just to not do it and to never suggest this binding... You decide.

I would second what @lolodomo is writing. The remote openHAB Binding is, from my perspective, targeting more advanced users, so there would be no real benefit from an auto suggestion.

david-pace reviewed (changes requested):
Please don't forget to remove the discovery method from the `boschshc` binding 😊

andrewfg commented:
> don't forget to remove the discovery method from the boschshc binding

AFAIK I did that already. ??

david-pace commented:
Hmm, I still see the changed `addon.xml` [here](https://github.com/openhab/openhab-addons/pull/15817/files#diff-3ea3a3b324bfcf1e444176dcb5c15fb8b6f90957f4f21d82046465a130d4a7b0). Can you also still see it there?

andrewfg commented:
> I still see the changed

@david-pace Ok done

david-pace reviewed (approved):
LGTM for `boschshc`

mvalla commented:
@andrewfg you may want to include openwebnet binding here? It uses UPnP to discover gateways as bridge Things.
It should be enough to check if manufactures ha “BTICINO” in the string.

See: https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/BusGatewayUpnpDiscovery.java

mvalla commented on bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -8,4 +8,17 @@
+			<match-properties>
+				<match-property>
+					<name>manufacturer</name>
+					<regex>(?i).*BTICINO.*</regex>
```
From the logs, actual strings that are found are:

`BTicino S.p.A.`
or
`BTicino S.p.A. (http://www.bticino.it)`
If the (?i) part of the regex ignores the case, the this should work

lolodomo commented:
@kaikreuzer : I will review this one. But it can't be merged before core stuff.

kaikreuzer commented:
Thanks @lolodomo! I plan to merge https://github.com/openhab/openhab-core/pull/3865 asap and then update the xsds on the website. I guess this should then be enough for this PR to correctly build and to be mergeable.

andrewfg commented:
> update the xsds on the website.
> this should then be enough for this PR to correctly build

I think so. (It does build already with -DskipChecks)

andrewfg commented:
@lolodomo just to report that @kaikreuzer merged the core dependency, so this PR does indeed now build properly. And I just synchronized this PR with openhab-addons:main so everything should now be clean.

lolodomo commented on bundles/org.openhab.binding.androidtv/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -8,4 +8,27 @@
+			<mdns-service-type>_nv_shield_remote._tcp.local.</mdns-service-type>
+			<match-properties>
+				<match-property>
+					<name>SERVER</name>
+					<regex>^$</regex>
```
You expect an empty value for the property SERVER ?
`.*` would be less restrictive.

lolodomo replied:
Should you also change the same way for androiddebugbtidge ?

lolodomo reviewed (commented):
Just one question

lolodomo reviewed (approved):
LGTM
I have of course not checked that the criteria are correct for all bindings.

lolodomo commented:
Thank you @andrewfg  (and @mherwege ) for all your efforts to introduce this new great feature (I mean this PR but also all the others). Even if it is useless for my own usage, it should help a lot of new OH users.
